CIP-DappConnector/CIP-dapp-connector.md

```diff
@@ -0,0 +1,226 @@
+
+### cardano.get_used_addresses(paginate: Paginate = undefined): Promise\<cbor\<address>[]>
+
+Errors: `APIError`
+
+Returns a list of all used (included in some on-chain transaction) addresses controlled by the wallet. The results can be further paginated by `paginate` if it is not `undefined`.
+
+### cardano.get_unused_addresses(): Promise\<cbor\<address>[]>
+
+Errors: `APIError`
+
+Returns a list of unused addresses controlled by the wallet.
+
+### cardano.get_change_address(): Promise\<cbor\<address>>
+
+Errors: `APIError`
+
+Returns an address owned by the wallet that should be used as a change address to return leftover assets during transaction creation back to the connected wallet. This can be used as a generic receive address as well.
+
+### cardano.sign_tx(tx: cbor\<transaction_body>, metadata: cbor\<auxiliary_data> = undefined): Promise\<cbor\<transaction>>
+
+Errors: `APIError`, `TxSignError`
+
+Requests that a user sign the supplied transaction body. The wallet should ask the user for permission, and if given, try to sign the supplied body and return a signed transaction. If the wallet could not sign the transaction, `TxSignError` shall be returned with the `ProofGeneration` code. Likewise if the user declined it shall return the `UserDeclined` code.
+
+### cardano.sign_tx_input(tx: cbor\<transaction_body>, index: number): Promise\<cbor\<transaction_witness_set>>
+
+Errors: `APIError`, `TxSignError`
+
+Provides lower-level ability for signing that produces the witnesses for just a single input in a transaction. This exists in case dApps need to construct a transaction to satisfy certain properties and the user might only own some of the inputs. The wallet should ask user permission for signing similar to `cardano.sign_tx_input()` and errors are handled in the same way.
+
+### cardano.sign_data(addr: cbor\<address>, sig_structure: cbor\<Sig_structure>): Promise\<Bytes>
+
+Errors: `APIError`, `DataSignError`
+
+This endpoint utilizes the [CIP-0008 signing spec](https://github.com/cardano-foundation/CIPs/blob/master/CIP-0008/CIP-0008.md) for standardization/safety reasons. It allows the dApp to request the user to sign data conforming to said spec. The user's consent should be requested and the details of `sig_structure` shown to them in an informative way. The  Please refer to the CIP-0008 spec for details on how to construct the sig structure.
+
+### cardano.submit_tx(tx: cbor\<transaction>): Promise\<hash32>
+
+Errors: `APIError`, `TxSendError`
+
+As wallets should already have this ability, we allow dApps to request that a transaction be sent through it. If the wallet accepts the transaction and tries to send it, it shall return the transaction id for the dApp to track. The wallet is free to return the `TxSendError` with code `Refused` if they do not wish to send it, or `Failure` if there was an error in sending it (e.g. preliminary checks failed on signatures).
+
```

Comment:
Probably we need some way to have a versioned number of this specification. There is already one wallet (Nami Wallet) that implements this specification, but clearly this spec will continue to change (esp. with the Alonzo specification)

We have to think about the fact not only may the user have multiple wallets that inject code, but also they may support different versions of this spec.
Probably the dapp will have to be able to say something like ask for any dApp connector the user has installed that implements API version `~1.x.x` or some other semver string.

Reply:
I updated the CIP so that every wallet that implements this spec has an object that has functions for connection + name (to show in the connect UI in the dApp) + a version. The dApp could then filter out the ones with incompatible versions when it's creating the "Connect to X wallet" buttons to initiate the dApp.

Reply:
We could also standardize a method on the `cardano` object which does that version filtering for you, which the first wallet to create that `cardano` object would have to inject into it too. Like `cardano.getWallets(versionSemVar: String): Wallet[]`. Otherwise it could just be made into a tiny external utility library instead to keep injected code to a minimum.